Fix bike check in delete

allenmichael099 opened the pull request:
Switched to object ID to identify which Checkinout entry to delete. It used to use user_id, but that did not delete database entries because it was not converted to BSON before querying. I couldn't find a way to convert it to BSON or a way to query with BSON in mongolite.

allenmichael099 committed:
Switched to object ID to identify entry to delete

allenmichael099 committed:
Store the object ID in the table

asiripanich commented on R/mod_DTedit.R:
```diff
@@ -73,12 +73,14 @@ mod_DTedit_server <- function(input, output, session, table_data, table_type,
     thedata = table_data,
     edit.cols = edit_columns,
-    view.cols = names(table_data),
+    view.cols = column_names[!column_names %in% '_id'], # view the table without the '_id' column
```
'_id' is a column name right? not a suffix?

allenmichael099 replied:
Right, '_id' is a column name here. I think it would be fine to wait until you've properly reviewed it.

asiripanich reviewed (commented):
Is this urgent? if not I would like to properly review it before we merge?

allenmichael099 committed:
Get target indices right

allenmichael099 committed:
Hide objectId from esquisse interactive plot

applied dplyr::select(-dplyr::any_of('_id')) %>%
to data_esquisse$data in app_server

allenmichael099 commented:
@shankari @asiripanich As far as I can tell, this is now ready to go. Checkinout objects get deleted based on objectId, and the formatted time still displays correctly.  

If in future edits to this feature you find the problem below, it is probably because the target indices within columnDefs got shifted out of place.
![Error_no_bike_data](https://user-images.githubusercontent.com/56986329/128062307-dfdf6f13-71cc-4ce4-aaca-0cfcee6beda1.PNG)

shankari reviewed (approved):
I think that the PR looks fine from an e-mission data model perspective, but I am not an R expert, and I don't know whether this is properly R-ic.

asiripanich commented:
> this feature you find the problem below, it is probably because the target indices

I personally don't like this workaround. Maybe DTedit is a bad choice to begin with. However, I'm going to merge it to allow this feature to exist. But it will most likely be removed when I do a major update to the dashboard.

asiripanich commented on R/mod_DTedit.R:
```diff
@@ -28,12 +28,18 @@ mod_DTedit_server <- function(input, output, session, table_data, table_type,
   if ("columnDefs" %in% names(DT_options)) {
-    DT_options$columnDefs[[1]]$targets <- DT_options$columnDefs[[1]]$targets - 1
-  }
+    # Adjust for id and for fmt_time
```
I put `browser()` inside this but I never stopped here. When do we need this?

allenmichael099 replied:
This is for when we need to render a formatted time. Currently it only is needed for Bike Check In, which has a times stamp that we want to also display as a human readable time in a separate column. I'm surprised the browser() didn't stop - it stopped there for me just now.

asiripanich replied:
Can you check that the test database you have is the same as the one in the repo?

allenmichael099 replied:
@asiripanich There probably was a difference, with mine containing a timestamp field and yours not. @shankari can provide you with data that has timestamps.

shankari commented:
@asiripanich I don't know that the problem is with DTedit; the problem is with making the call to the database to delete the entry. We will have to solve something similar for the user status editing anyway.

asiripanich reviewed (commented):
@allenmichael099

asiripanich commented:
@allenmichael099 could you please explain why you need this? I know it is not part of this PR but seems related to this.

https://github.com/asiripanich/emdash/blob/e44b82c3be282d22e2ca0e11966116533c568028/R/app_server.R#L202-L217

allenmichael099 commented:
@asiripanich The code snippet above is so we will render a human readable time next to the timestamp every time a supplementary table has a column called 'ts'.

asiripanich commented:
> 
> 
> @asiripanich The code snippet above is so we will render a human readable time next to the timestamp every time a supplementary table has a column called 'ts'.

Can we convert this timestamp to a proper datetime field during the 'load data' step?

shankari commented:
@asiripanich we tried using datetime, but since the conversion happens in R, it converts based on the timezone of the server, which for AWS is UTC. So instead of the user seeing the datetime in their local timezone, they will see it in UTC, and have to do the conversion in their head.

With this option, we convert the timestamp on the browser instead.

asiripanich committed:
Merge branch 'master' into Fix_bike_checkin_delete
